feat: add Agent

julian-risch opened the pull request:
### Related Issues
- fixes #4100

### Proposed Changes:
- Added new Agent node according to the [proposal ](https://github.com/deepset-ai/haystack/blob/main/proposals/text/3925-mrkl-agent.md)
- Added an AgentError class for all errors raised related to Agent
- Added new tests for the agent as unit tests and integration tests
- Added a MockPromptNode for those tests and extended MockRetriever
- Updated labeler.yml workflow so that topic:agent is added to PRs with changes to agent related files
- Updated tests.yml workflow so that tests added for the agent are executed as unit tests and integration tests in the CI

### How did you test it?
- Added test cases 
- Manual tests for example with the following branches: 
  - https://github.com/vblagoje/haystack/tree/web_retriever_example
  - https://github.com/vblagoje/haystack/tree/agent_with_webretriever
- Checked that `Agent initialized` and `AgentError raised` events are correctly received by posthog

### Notes for the reviewer
- There is no `send_agent_event` implemented for Agent but we might want to have it in future as we also have the `send_pipeline_event` for Pipeline.
- In a previous version of this PR I included an example Agent yaml and skipped tests for saving to/loading from YAML. However, we might not implement save to/load from YAML for Agent's as planned in https://github.com/deepset-ai/haystack/issues/4099 in the next weeks, so I removed these to not confuse users.
- I did not compile the regex before using it because we use regexes only in two other parts of Haystack so the compiled regex should be cached automatically according to https://docs.python.org/3/library/re.html#re.compile

Known limitations:
- `params` passed to `agent.run()` will not be forwarded to tools that are nodes but only to tools that are pipelines. The reason is that `node.run()` does not accept params but `pipeline.run()` does.

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I have updated the related issue with new insights and changes
- [x] I added tests that demonstrate the correct behavior of the change
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,266 @@
+    """
+
+    def __init__(self, prompt_node: PromptNode, tools: Optional[List[Tool]] = None, max_iterations: int = 10):
+        self.prompt_node = prompt_node
```
Should we even allow agents without tools?

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,271 @@
+
+            prompt_template = PromptTemplate("think-step-by-step", transcript, ["query"])
+            pred = self.prompt_node.prompt(prompt_template=prompt_template, query=query)
+            transcript += f"{pred[0]}\n"
```
Let's add function `_extract_prediction` - it returns string that we assign to a variable here. So we'll have `pn_pred = self._extract_prediction(pred)` and then we can use pn_pred as a str wherever we need to, including on lines 154 and 156. We thus enable different prediction extraction is subclasses if needed and clean up this code a bit at the same time.

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,271 @@
+                    f'The Agent tried to use a tool "{tool_name}" but the registered tools are only {self.tools.keys()}.'
+                )
+            pipeline_or_node = self.tools[tool_name].pipeline_or_node
+            # We can only pass params to pipelines but not to nodes
```
I would replace everything from 166 to 189 with method call `_extract_observation`, which has the parameter `tool_name`. Let's hide all this nitty-gritty logic from the main loop

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,271 @@
+                else:
+                    # no answer/document/result was returned
+                    observation = ""
+            transcript += f"Observation: {observation}\nThought: Now that I know that {tool_input} is {observation}, I "
```
Don't include 189 in `_extract_observation`; it's important imho to have this line of code here (for readability).

vblagoje reviewed (changes requested):
Looks great, I left some minor comments to make everything more readable and easier to follow for a complete newcomer to this code

julian-risch commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,271 @@
+        self.prompt_node = prompt_node
+        if tools is not None:
+            self.tools = {tool.name: tool for tool in tools}
+            self.prompt_text = self._generate_prompt_text()
```
Thank you so much for the initial feedback. It's all addressed now, except for adding `prompt_text` as an init parameter of the `Agent`. This one needs more discussion.

julian-risch replied:
I changed the init method so that users can pass a registered PromptTemplate in the Agent's init method. It just needs to be registered with the PromptNode beforehand, which should be no problem.

julian-risch commented:
@agnieszka-m I already added a link to https://docs.haystack.deepset.ai/docs/agents in the Agent's error messages. The page obviously doesn't exist yet but the URL would be consistent with our other documentation pages. Please just let me know if a different URL makes more sense and I'll change the error message.

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
```
Pythonic perfectionists will say that this is not pythonic and that we should use `if not self.tools:`

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+                )
+            tool_name, tool_input = self._extract_tool_name_and_tool_input(pred=pred)
+            if tool_name is None or tool_input is None:
+                raise AgentError("Wrong output format.")
```
Include a trace of pred into AgentError here?

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+
+            # Try to extract the final answer or a tool name and tool input from the generated text
+            final_answer = self._extract_final_answer(pred=pred)
+            if final_answer is not None:
```
Python perfectionists ask for `if final_answer:`

julian-risch replied:
If `final_answer` is `""` that evaluates to `False` so I would keep that as is.

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+            tool_names_with_descriptions=tool_names_with_descriptions,
+            generated_text=transcript,
+        )
+
```
We could invert the if statement on line 157. So if we don't have a final answer - we take a step. And else, if we do, we format that answer and return it. Not sure if possible, but if it is - could save us 6-7 lines of code in calling _format_answer?

julian-risch replied:
There is a third case, which needs to be handled if `max_iterations` is reached. So if I use a `break` condition it's enough to call `_format_answer` only once in the code, otherwise twice I think. Happy to discuss when we meet.

julian-risch replied:
@vblagoje I would leave the loop as it is now. The call to `_format_answer` is simpler but there are still two calls of that function. Could you please have another look to see whether you could approve the current version?

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+        self, tool_name: str, tool_input: str, params: Optional[dict] = None
+    ) -> Union[Tuple[Dict[str, Any], str], Dict[str, Any]]:
+        if not self.has_tool(tool_name):
+            raise AgentError(
```
Should we include transcripts in AgentError by default? It could be useful I think

julian-risch replied:
Would be useful I agree. 👍 I will check how to do that. One downside is that if the Agent generates text with `$` in it, filling the template will probably fail. And filling the template is required to have a proper transcript.

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+        if isinstance(result, tuple):
+            result = result[0]
+        if isinstance(result, dict):
+            if "results" in result and len(result["results"]) > 0:
```
Let's see if we need those len calls

vblagoje commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+            "query": query,
+            "answers": [Answer(answer=answer, type="generative")],
+            "transcript": next(
+                self.prompt_node.get_prompt_template(self.prompt_template).fill(
```
We have a reference to the template in `self.prompt_template`. Can we call it directly or it has to be via prompt node?

julian-risch replied:
`self.prompt_template` is just the name of the template. So to get the `think-step-by-step` template I think we always need the prompt node.

julian-risch replied:
I'll change the implementation so that `self.prompt_template` is a `PromptTemplate`. The function call will become `next(self.prompt_template.fill(` instead of `next(self.prompt_node.get_prompt_template(self.prompt_template).fill(` instead of

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -681,6 +681,27 @@ def get_predefined_prompt_templates() -> List[PromptTemplate]:
             prompt_text="Translate the following context to $target_language. Context: $documents; Translation:",
         ),
+        PromptTemplate(
+            name="think-step-by-step",
```
What if we call this template zero-shot-react? We could also provide a few-shot-react with examples. Let's talk offline.

julian-risch replied:
I have no strong opinion on the template name. Happy to discuss alternatives. 👍

julian-risch replied:
We decided to call this template `zero-shot-react`. In another PR, we could later add a template `few-shot-react`.

vblagoje replied:
ReAct to honour the ideas from https://arxiv.org/abs/2210.03629

silvanocerza commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+        """
+        self.prompt_node = prompt_node
+        self.prompt_template = prompt_template
+        self.tools = {tool.name: tool for tool in tools} if tools else {}
```
Won't this fail if `tools` is `None`?

silvanocerza commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+
+        :param query: The search query.
+        :param max_iterations: The number of times the Agent can run a tool plus then once infer it knows the final answer.
+            Set at least to 2 so that the Agent can run one tool and then infer it knows the final answer. Default 5.
```
Isn't this a bit misleading? Technically the default is whatever has been set in the `Agent` constructor, if we set it to 10 in there but leave it untouched in `run` it will default to 10, right? 🤔

julian-risch replied:
The idea is that we might want to use different values of max_iterations depending on the query but all with the same Agent. So a user can set a default when initializing the Agent and then later overwrite it.

julian-risch replied:
Ah, you mean that `Set at least to 2` is confusing because it's already set in the Agent's init. 👍 I agree.

julian-risch replied:
I updated the docstring to make clear that the parameter is optional and removed the statement about the default value.

silvanocerza commented on haystack/agents/base.py:
```diff
@@ -0,0 +1,298 @@
+
+        pipeline_or_node = self.tools[tool_name].pipeline_or_node
+        # We can only pass params to pipelines but not to nodes
+        if isinstance(pipeline_or_node, (Pipeline, BaseStandardPipeline)):
+            result = pipeline_or_node.run(query=tool_input, params=params)
+        else:
+            if isinstance(pipeline_or_node, BaseRetriever):
+                result = pipeline_or_node.run(query=tool_input, root_node="Query")
+            else:
+                result = pipeline_or_node.run(query=tool_input)
```
```suggestion
        if isinstance(pipeline_or_node, (Pipeline, BaseStandardPipeline)):
            result = pipeline_or_node.run(query=tool_input, params=params)
        elif isinstance(pipeline_or_node, BaseRetriever):
            result = pipeline_or_node.run(query=tool_input, root_node="Query")
        else:
            result = pipeline_or_node.run(query=tool_input)
```

silvanocerza commented on haystack/agents/base.py:
```diff
@@ -129,60 +139,39 @@ def run(
-            )[0]
-
-            # Try to extract the final answer or a tool name and tool input from the generated text
+            pred = self.prompt_node(transcript)[0]
```
This will fail if an empty list is returned, am not sure how likely that is but I would raise a clear exception.

silvanocerza commented on test/agents/test_agent.py:
```diff
@@ -0,0 +1,249 @@
+    assert "Maximum number of iterations (1) reached" in caplog.text
+
+    # Setting max_iterations in the Agent's run method
+    with caplog.at_level(logging.WARN, logger="haystack.agents"):
+        result = agent.run("Where does Christelle live?", max_iterations=0)
+    assert result["answers"] == [Answer(answer="", type="generative")]
+    assert "Maximum number of iterations (0) reached" in caplog.text
```
Maybe we should raise an exception when a user sets `max_iterations` to 0? I don't see the reason why that should be possible?

julian-risch replied:
There is no good reason for setting `max_iterations` to 0 or 1, that's correct. I will change the test to use 3 and 2 instead of 1 and 0. Not sure whether it's worth checking and raising an exception though. I'll need to think about it.

julian-risch replied:
We'll raise an error now if `max_iterations` is smaller than 2.

silvanocerza commented on test/agents/test_agent.py:
```diff
@@ -0,0 +1,249 @@
+        )
+    )
+
+    result = agent.run("How many characters are in the word Madrid?")
+    assert any(digit in result["answers"][0].answer for digit in ["5", "6", "five", "six"])
```
Why do we include `5` and `five` here? Won't the answer always be `len("Madrid")`?

julian-risch replied:
I had an off by one error once and it would be a shame if we need to rerun the tests because of that.

silvanocerza commented on test/agents/test_agent.py:
```diff
@@ -0,0 +1,249 @@
+    result = agent.run("How many characters are in the word Madrid?")
+    assert any(digit in result["answers"][0].answer for digit in ["5", "6", "five", "six"])
+
+    result = agent.run("How many letters does the name of the town where Christelle lives have?")
+    assert any(digit in result["answers"][0].answer for digit in ["5", "6", "five", "six"])
```
Same for this one, not sure who Christelle is but I'd expect the name of a town to be consistent. 😅

julian-risch replied:
This `Counter` tool is probably not well suited for the tests. Counting characters doesn't work well with the LLMs it seems. Probably because they work on token level. I'll see whether I find a better tool for the tests. Once some tools are implemented in Haystack, we can use them here. Consider it a placeholder until we have proper tools.

silvanocerza replied:
Good for me then. We can rework that in the future. 👍

silvanocerza reviewed (changes requested):
Raised some concerns but overall looks good.

julian-risch commented:
> Raised some concerns but overall looks good.

@silvanocerza Thank you so much for reviewing this large PR. I think I addressed your remaining concerns now by:
* Raising an error if max_iterations<2
* Raising an error if preds returned from prompt node are an empty list
* Adding a TODO into tests so that we replace the Count tool once we have proper tools implemented as planned in https://github.com/deepset-ai/haystack/pull/4085
* Updating the docstring in the run method so that no default is mentioned and it is more clear that the value is optional

vblagoje reviewed (approved):
Much improved from the early experiment, looks very solid now!

silvanocerza reviewed (approved):
🚀
